[WIP] Terraform Docs for TF policy (#2690)
hashicorp/web-unified-docs: content/terraform-policy/v0.1.x (beta)/docs/policy/reference/policy/provider-policy.mdx

---
page_title: `provider_policy`
description: >-
  Reference for the `provider_policy {}` block used to write policies for Terraform providers.
---

# `provider_policy`

The `provider_policy` block defines policies for Terraform providers. The first label of the block identifies the provider name, such as `aws`, `google`, or `azurerm`, and the second label is the policy name. You can use a wildcard (*) for the provider name to apply a single policy to every provider defined in the configuration.

@include "partials/beta.mdx"

## Configuration model

```hcl
provider_policy "<provider_name>" "<policy_name>" {

>   operation         = "<operation>"

Reviewer: operation is not valid attribute for provider policy

  enforcement_level = "<level>"
  filter            = <boolean_expression>

  locals {
    name = <value>
  }

  enforce {
    condition     = <boolean_expression>
    error_message = "<message>"
    info_message  = "<message>"
  }

  enforce { ... }
}
```

## Complete configuration example

```hcl
provider_policy "aws" "standard_configuration" {
  operation         = ["create", "update"]
  enforcement_level = "mandatory"

  locals {
    version_constraint = core::semverconstraint(meta.version, ">= 5.0.0")
  }

  enforce {
    condition     = attrs.region == "us-east-1"
    error_message = "All AWS resources must be deployed in the us-east-1 region."
  }

  enforce {
    condition     = meta.source == "hashicorp/aws"
    error_message = "Only the official HashiCorp AWS provider is permitted."
  }

  enforce {
    condition     = local.version_constraint
    error_message = "The AWS provider version must be at least 5.0.0."
  }
}
```

## Specification

### Labels

| Label | Required | Description |
|-------|----------|-------------|
| First label | Yes | The provider name (For example: `aws`, `google`, `azurerm`). Use `*` to match all providers. |

> | Second label | Yes | A unique name for this policy within the policy file. |

Reviewer 2: ```suggestion

The `provider_policy` block supports the following configuration.

### `provider_policy "<TYPE>" "<LABEL>"`

You must set the following arguments in the `provider_policy` block:

- `TYPE`: Specifies the provider name, such as `aws`, `google`, `azurerm`. You can use the `*` wildcard to match all providers. 
- `LABEL`: Specifies a unique name for this policy within the policy file. 
```
Have a look at other reference docs for prior art: https://developer.hashicorp.com/terraform/language/block/data#specification

### Arguments

Provider policy blocks support the following arguments.

| Argument | Required | Type | Default | Description |
|----------|----------|------|---------|-------------|

> | `operation` | No | List of Strings | `["create", "update"]` | The operations during which to enforce this policy. Valid values: `"create"`, `"update"`, `"delete"`. |

Reviewer: Operation attribute is not valid for provider and module policy and need to be removed
cc: @vpaul97 @viveksinghgit95

> | `enforcement_level` | No | String | `"advisory"` | The enforcement level for the policy. Valid values: `"advisory"`, `"mandatory"`, `"mandatory-overridable"`. |

Reviewer: ```suggestion
| `enforcement_level` | No | String | `"advisory"` | The enforcement level for the policy. Valid values: `"advisory"`, `"mandatory"`, `"mandatory_overridable"`. |
```

> | `filter` | No | Boolean expression | - | Filter which providers the policy applies to based on their attributes. |

Reviewer 2: ```suggestion
### `enforcement_level`

Specifies the enforcement level for the policy. You can specify one of the following values:

- `"advisory"`: Allows the operation to continue but prints a warning when if the conditions of the policy aren't met. 
- `"mandatory"`: Stops the operation if the conditions of the policy aren't met.
- `"mandatory_overridable"`: Stops the operation if the conditions of the policy aren't met, but allows a user with appropriate permissions to override the policy and continue. 

 Refer to []() for more information about policy enforcement levels.
 
  - Data type: String
  - Default: None 
   

### `filter` 

Specifies a Boolean expression that filters which provider configurations the policy applies to based on their attributes. The left side of the expression references an attribute in the provider configuration. The right side specifies an attribute value. When the expression evaluates to `true`, the policy applies to the provider. Refer to [Filter by provider alias](#filter-by-provider-alias) for an example.

You can prefix the left side of the expression with the `meta` object. This object allows you to select a Terraform-specific argument in the provider configuration, such as `source`, `version`, and `alias`. Refer to [`meta` object attributes](#meta-object-attributes) for more information.  

You can also use the `attrs` object in your expression to access arguments defined in the provider block, such as region settings or account identifies. The attributes you can select with the `attrs` depend on the provider. 

- Data type: Boolean expression
- Default: None

```
I _think_ this is how it works?

### Attributes

Within the `provider_policy {}` block, you can reference the following attirubites.

| Object | Description |
|--------|-------------|
| `attrs` | Provides access to the arguments defined in the provider block, such as region settings or account identifiers. |

> | `meta` | Provides access to Terraform-defined information about the provider, including `source`, `version`, and `alias`. |

Reviewer 2: ```suggestion
```
I _think_ we can replace this section with an explanation of the `attrs` selector. One big reason for the configuration model is to provide the hierarchical overview so that we can explain elements in the configuration individually, as opposed to adding subsections, such as "Attributes", because it becomes difficult to understand the relationships between them.

#### Arguments

Enforce blocks support the following arguments.

| Argument | Required | Type | Description |
|----------|----------|------|-------------|
| `condition` | Yes | Boolean expression | A boolean expression that must evaluate to `true` for the policy to pass. |
| `error_message` | Yes | String | The error message to display when the condition evaluates to `false`. |
| `info_message` | No | String | An informational message to display regardless of whether the policy passes or fails. |

You can include one or more `enforce {}` blocks in each policy block. If any of the enforce blocks' conditions fail during evaluation, then Terraform policy will fail the policy and return the corresponding error messages.

### The `locals {}` block

Each `provider_validation_policy {}` block may contain `locals {}` blocks that define values that are local to the policy. These local values override `locals {}` blocks defined at the top level of the policy file. Refer to [The `locals {}` block](/terraform/policy/reference/policy/locals) for more information.

## Examples

> 

Reviewer 2: ```suggestion
The following examples show how to write configuration for common use cases.
```

### Enforce provider source and version

> 

Reviewer 2: ```suggestion
The following policy includes a condition that sets the minimum version of the provider to 5.0.0. The condition accesses provider version using the `meta.version` selector and passes it to the `core::semverconstraint` function.
```
I'm not 100% sure if this explanation is accurate, but ideally we'd introduce each example by describing the salient parts of the configuration. We might have to add these later.

  }

  enforce {
    condition     = meta.source == "hashicorp/aws"
    error_message = "Only the official HashiCorp AWS provider is permitted."
  }

  enforce {
    condition     = local.version_constraint
    error_message = "The AWS provider version must be at least 5.0.0."
  }
}
```

## Specification

The `provider_policy` block supports the following configuration.

### `provider_policy "<TYPE>" "<LABEL>"`

You must set the following arguments in the `provider_policy` block:

- `TYPE`: Specifies the provider name, such as `aws`, `google`, `azurerm`. You can use the `*` wildcard to match all providers. 
- `LABEL`: Specifies a unique name for this policy within the policy file. 
### `enforcement_level`

Specifies the enforcement level for the policy. You can specify one of the following values:

- `"advisory"`: Allows the operation to continue but prints a warning when if the conditions of the policy aren't met. 
- `"mandatory"`: Stops the operation if the conditions of the policy aren't met.
- `"mandatory_overridable"`: Stops the operation if the conditions of the policy aren't met, but allows a user with appropriate permissions to override the policy and continue. 

>  Refer to []() for more information about policy enforcement levels.

Reviewer: will this link be added later ?
